src/client/index.js

```diff
@@ -1,137 +1,99 @@
-          
-          // If we aren't already connected then throw an error
-          if (!socket.connected) {
-            throw new Error('Socket not connected');
-          }
-          
-          socket.emit('authenticate', options);
+        if (options.type === 'local') {
+          endPoint = config.localEndpoint;
+        } else if (options.type === 'token') {
+          endPoint = config.tokenEndpoint;
+        } else {
+          throw new Error(`Unsupported authentication 'type': ${options.type}`);
         }
 
-        // If we are using primus
-        if (app.primus) {
-          const socket = app.primus;
-          
-          // If we aren't already connected then throw an error
-          if (socket.readyState !== 3) {
-            throw new Error('Socket not connected');
+        return connected(app).then(socket => {
+          // TODO (EK): Handle OAuth logins
+          // If we are using a REST client
+          if (app.rest) {
+            return app.service(endPoint).create(options).then(handleResponse);
           }
-          
-          socket.send('authenticate', options);
-        }
+
+          const method = app.io ? 'emit' : 'send';
+
+          return authenticateSocket(options, socket, method).then(handleResponse);
+        });
       });
     };
 
-    app.user = function() {
-      return storage().get('user').then(data => data.value);
-    };
-    
     app.token = function() {
-      return storage().get('token').then(data => data.value);
+      return getJWT(config.tokenKey, this.get('storage'));
     };
 
     app.logout = function() {
-      return storage().remove(null, { id: { $in: ['user', 'token' ] } });
+      app.set('user', null);
+      app.set('token', null);
+      // TODO invalidate token
+      return Promise.resolve(app.get('storage').setItem(config.tokenKey, null));
     };
 
     // Set up hook that adds adds token and user to params so that
     // it they can be accessed by client side hooks and services
     app.mixins.push(function(service) {
-      if (!service.before || !service.after) {
-        throw new Error(`It looks like feathers-hooks isn't configured. It is required before you configure feathers-authentication.`);
+      if (typeof service.before !== 'function' || typeof service.after !== 'function') {
+        throw new Error(`It looks like feathers-hooks isn't configured. It is required before running feathers-authentication.`);
       }
-      service.before(hooks.populateParams(authOptions));
+
+      service.before(hooks.populateParams(config));
     });
-    
+
     // Set up hook that adds authorization header for REST provider
     if (app.rest) {
       app.mixins.push(function(service) {
-        if (!service.before || !service.after) {
-          throw new Error(`It looks like feathers-hooks isn't configured. It is required before you configure feathers-authentication.`);
+        if (typeof service.before !== 'function' || typeof service.after !== 'function') {
```

Comment:
I realized that we probably don't need this second duplicate warning because it should be caught by the check above.